src/devices/ikea_sparsnas.c

```diff
@@ -0,0 +1,251 @@
+    
+    for (size_t i = 0; i < 5; i++)
+        decrypted[i] = buffer[i];
+
+    for(size_t i = 0; i < 13; i++)
+        decrypted[5 + i] = buffer[5 + i] ^ key[i % 5];
+
+    if (decoder->verbose > 1)
+        decoder_output_bitrowf(decoder, decrypted, 18 * 8, "Decrypted");
+
+    // Additional integrity checks
+
+    int rcv_sensor_id = decrypted[5] << 24 | decrypted[6] << 16 | decrypted[7] << 8 | decrypted[8];
+
+    if (rcv_sensor_id != IKEA_SPARSNAS_SENSOR_ID) {
+        if (decoder->verbose > 1)
+            fprintf(stderr, "IKEA Sparsnäs: Received sensor id (%d) not the same as sender (%d)\n", rcv_sensor_id, IKEA_SPARSNAS_SENSOR_ID);
+        return 0;
+    }
+
+    if (decrypted[0] != 0x11){
+        decoder_output_bitrowf(decoder, decrypted + 5, 13 * 8,  "Message malformed");
+        if (decoder->verbose > 1)
+            fprintf(stderr, "IKEA Sparsnäs: Message malformed (byte0=%X expected %X)\n", decrypted[0], 0x11);
+        return 0;
+    }
+    
+    if (decrypted[3] != 0x07){
+        decoder_output_bitrowf(decoder, decrypted + 5, 13 * 8,  "Message malformed");
+        if (decoder->verbose > 1)
+            fprintf(stderr, "IKEA Sparsnäs: Message malformed (byte3=%X expected %X)\n", decrypted[0], 0x07);
+        return 0;
+    }
+
+    if (decoder->verbose > 1)
+        fprintf(stderr, "IKEA Sparsnäs: Received sensor id: %d\n", rcv_sensor_id);
+    
+    //Value extraction and interpretation
+
+    uint16_t sequence_number = (decrypted[9] << 8 | decrypted[10]);
+    uint16_t effect = (decrypted[11] << 8 | decrypted[12]);
+    uint32_t pulses = (decrypted[13] << 24 | decrypted[14] << 16 | decrypted[15] << 8 | decrypted[16]);
+    uint8_t battery = decrypted[17];
+    double watt = effect * 24;
+    uint8_t mode = decrypted[4]^0x0f;
+
+    //Note that mode cycles between 0-3 when you first put in the batteries in
+    if(mode == 1){
+      watt = (double)((3600000 / IKEA_SPARSNAS_PULSES_PER_KWH) * 1024) / (effect);
+    } else if (mode == 0 ) { // special mode for low power usage
+      watt = effect * 0.24 / IKEA_SPARSNAS_PULSES_PER_KWH;
+    }
+    
+    double cumulative_kWh = ((double)pulses) / ((double)IKEA_SPARSNAS_PULSES_PER_KWH);
+    
+
+    data_t *data;
+    data = data_make(
+        "model",         "Model",               DATA_STRING, "IKEA Sparsnäs Energy Meter Monitor",
+        "id",            "Sensor ID",           DATA_INT, rcv_sensor_id,
+        "sequence",      "Sequence Number",     DATA_INT, sequence_number,
+        "battery",       "Battery",             DATA_FORMAT, "%d%%", DATA_INT, battery,
+        "cumulative_kWh", "Cumulative kWh",     DATA_FORMAT, "%7.3fkWh", DATA_DOUBLE,  cumulative_kWh,
+        "effect",        "Effect",              DATA_FORMAT, "%dW", DATA_INT,  effect,
+        "pulses",        "Pulses",              DATA_INT,  pulses,
+        "mode",          "Mode",                DATA_INT, mode,
```

Comment:
Add mic = CRC here. Look in other decoders that have mic.

Reply:
OK